### ISSUE TITLE
A cached Session/KV path for gemma4 — the forward is cacheless, so every turn re-prefills the whole prompt

### ISSUE THREAD

**Issue by anthony-saronic** [anthony-chaudhary/fak#5496]
**Title:** `feat(model): a cached Session/KV path for gemma4 — the forward is cacheless, so every turn re-prefills the whole prompt`

**Labels:** `enhancement`, `model-arch`, `pagedattention`, `track/A-model-support`, `class:dev`

## The claim

The gemma4 forward has **no cached Session path**. The diagnostic flag that exists to work around
that says so in its own help text — `cmd/diagtok/main.go:26`:

> `--cacheless` … "greedy-decode via repeated cacheless Forward (for arches without a cached Session
> path yet, e.g. gemma4)"

Consequences, in order of cost:

1. **Decode re-runs the sequence forward per token** instead of extending a cache.
2. **There is no prefix cache**, so none of the reuse machinery (`internal/radixkv`,
   `internal/l3kv`, `internal/kvmmu`, `internal/cachemeta`) can engage for this family. A
   shared-prefix or multi-agent story is unavailable on this arch by construction.
3. An agent loop re-prefills its entire standing head (system prompt + tool schemas + context)
   **every turn, for every concurrent session** — which on a self-hosted single-GPU deployment is
   the dominant cost of a turn, ahead of decode.

## ⭐ Why this arch is an unusually good target, not a hard one

From the public `config.json` of `google/gemma-4-26B-A4B-it` (Apache-2.0), verifiable by anyone:

- `text_config.num_hidden_layers: 30`, and `layer_types` is a repeating **five `sliding_attention`
  then one `full_attention`** pattern ⇒ **25 sliding + 5 global**.
- `text_config.sliding_window: 1024` — a hard cap, not a hint.
- `head_dim: 256` but `global_head_dim: 512` — the two regimes have **different per-layer geometry**,
  which is why a dedicated gemma4 forward exists at all.

A sliding layer at position *t* never attends further back than 1024 tokens. So the state sufficient
to resume **any** continuation of a cached prefix is exactly:

```
5 global layers  × full prefix depth
+ 25 sliding layers × the last 1024 tokens
```

**The sliding tier stops growing at the window.** Five-sixths of the layers contribute a *constant*
to a persisted prefix regardless of how long that prefix is — so a cached prefix on this family is a
**bounded object**, where a uniformly-global model's grows with context. That is an unusually cheap
target for prefix persistence and reuse.

The information the cache needs is already parsed: `applyGemma4Config`
(`internal/ggufload/gguf_config.go:317`) fills `Config.Window` per layer from the header's
`attention.sliding_window_pattern`, alongside `HeadDimPerLayer` and `NumKVHeadsPerLayer`.

## Definition of done

- [ ] A cached `Session` for heterogeneous window/global geometry: per-layer KV extents driven by
      `Config.Window`, not one uniform buffer.
- [ ] Window-capped layers **bound** their KV allocation at the window rather than at the context
      length.
- [ ] A prefix-reuse test: a second turn over the same prefix prefills only the delta, and its logits
      match the cacheless forward within tolerance.
- [ ] `--cacheless` stops being the only way to decode this arch (keep the flag as a reference-path
      diagnostic).

**Comment by anthony-chaudhary** [comment]
Scouted this against the local trunk (currently 51 commits ahead of `origin/main`, so GitHub's view of this ticket is stale by exactly the commits that shrink it). Three corrections and a re-scoping.

## Two DoD bullets are already satisfied

`d6c8a1f10` (#5495) wired the dedicated gemma4 forward onto the resident-quant Session path. `Session.Prefill` (`internal/model/kv.go:642-648`), `Step` (`:862-864`) and `PrefillNoLogits` all dispatch gemma4 today, to `prefillGemma4`/`stepGemma4` (`internal/model/gemma4_session.go:115-123`).

So:

- **"`--cacheless` stops being the only way to decode this arch" is DONE.** It has not been the only way since `d6c8a1f10`. That bullet should be re-scoped to "…in O(n) time".
- **"the forward has no cached Session path" is half-true and worth restating precisely.** There *is* a Session path; it is a **recompute** session, documented as such at `gemma4_session.go:28-34`. `gemma4StackLast` re-runs `m.layerGemma4` over the entire retained history on every ingest, so it is O(n²) — correct, not cheap.
- The `cmd/diagtok/main.go:26` help text this ticket quotes as evidence ("for arches without a cached Session path yet, e.g. gemma4") is itself now stale prose, and is a free byproduct of any work here.

## The window is predicted and masked, but not bounded — and the repo already says so

Worth recording because it is the crux of DoD bullet 2. `internal/model/contextsize.go:24-32` is an explicit in-repo admission that the discount is "only TRUE if the cache actually drops aged-out positions on a windowed layer, **and today it does not**".

Confirming that from three directions:

- `Session.TrimToWindow` (`internal/model/swa.go:48`) has no production caller.
- `Config.MaxWindow` (`swa.go:25-37`) returns 0 unless EVERY layer declares a window — and gemma4's global layers get `Window[l] = -1` (`internal/ggufload/gguf_config.go:404`), so it is **structurally 0 for every gemma4 checkpoint**.
- `internal/compute/capacity.go:130-152 windowCappedPositions` feeds only `EstimateKVStoreBytes` — a byte-count predictor that never touches a live buffer. `80ee25b216` (#5520) wired it in behind `FAK_HYBRID_KV`, off by default.

The one place the window bites at runtime is a **read mask, not an allocation bound**: `internal/model/kv.go:540` `windowLoStep(...)` skips out-of-window rows in the dot product, but they are still stored. Today: window predicted ✔, masked ✔, **bounded ✘**.

## DoD bullet 2 is epic-sized, and should be split out

Bounding allocation per layer is blocked on `KVCache.pos` being **one array shared across all layers** (`internal/model/kvcache.go:17`). Under a real per-layer window, layer 5 holds 1024 rows while a global layer holds 40,000 — at which point `Cache.Len()` stops being well-defined, and every `s.Cache.Len()` caller in `Prefill`/`Step`/`PrefillNoLogits` (`kv.go:736,742,756,834,867,872`) plus the in-kernel decode path breaks. That is a refactor of the cache's central invariant, not a rung of this ticket.

The rest of the geometry is friendlier than that makes it sound: `KVCache.K` is `[][]float32` (`kvcache.go:13`), i.e. already per-layer, so a layer *can* hold rows of width `numKVHeadsForLayer(l) * headDimForLayer(l)`. Only `kvStride()` (`kvcache.go:42`) and `evictSupported` (`:126-131`) assume uniformity, and neither sits on the append/read path.

## Suggested split

1. **#5548** — a correctness bug that exists *because* this ticket has not landed: gemma4 partial prefix reuse prefills only the suffix. Filed separately; should land first.
2. **`tokenHiddenGemma4`, a per-position cached step** — this ticket's bullets 1 and 4, landable alone with `Evict`/`Reserve` fail-closed by a typed verdict (the precedent is `RecurrentEvictUnsupportedError`, `kvcache.go:54-64`). The template is MiniMax-M3 MSA (`internal/model/minimax_m3_session.go`), which was built as exactly this kind of "re-derive the cacheless math one position at a time" bridge; copy its five-function shape and its witness `TestMiniMaxM3MSASessionCacheMatchesCacheless` (`internal/model/minimax_m3_test.go:340-399`). The gemma4 twin of its fail-closed guard already exists as `requireGemma4Session` (`gemma4_session.go:57-66`).
3. **Per-layer window-bounded allocation** — bullet 2, its own epic, gated on the `KVCache.pos` refactor above.

## One more thing, unrelated to the split

`compute.KVConfig` carries only `WindowPerLayer` (`internal/compute/compute.go:275-289`) — no per-layer head geometry. `EstimateKVStoreBytes` therefore computes `elemsPerRow` from the **scalar** `cfg.NumKVHeads * cfg.HeadDim` (`capacity.go:110`), and `applyGemma4Config` sets that scalar to `keyLenFull` — the **global**, larger value (`gguf_config.go:408-410`). So every sliding layer is charged at the global head width. It over-counts, which is the safe direction for an admission gate, but `internal/kvbudget/perlayer.go:57-64` warns that "a caller that sets one should set both", and `compute` does not.
